crates/bevy_tasks/src/task_pool.rs

```diff
@@ -162,58 +166,63 @@ impl TaskPool {
+                spawned: Vec::new(),
             };
 
-            // Pin the future on the stack.
-            pin!(fut);
+            f(&mut scope);
+
+            if scope.spawned.is_empty() {
+                Vec::default()
+            } else if scope.spawned.len() == 1 {
+                vec![future::block_on(&mut scope.spawned[0])]
+            } else {
+                let fut = async move {
+                    let mut results = Vec::with_capacity(scope.spawned.len());
+                    for task in scope.spawned {
+                        results.push(task.await);
+                    }
 
-            // SAFETY: This function blocks until all futures complete, so we do not read/write the
-            // data from futures outside of the 'scope lifetime. However, rust has no way of knowing
-            // this so we must convert to 'static here to appease the compiler as it is unable to
-            // validate safety.
-            let fut: Pin<&mut (dyn Future<Output = Vec<T>> + Send)> = fut;
-            let fut: Pin<&'static mut (dyn Future<Output = Vec<T>> + Send + 'static)> =
-                unsafe { mem::transmute(fut) };
-
-            // The thread that calls scope() will participate in driving tasks in the pool forward
-            // until the tasks that are spawned by this scope() call complete. (If the caller of scope()
-            // happens to be a thread in this thread pool, and we only have one thread in the pool, then
-            // simply calling future::block_on(spawned) would deadlock.)
-            let mut spawned = self.executor.spawn(fut);
-            loop {
-                if let Some(result) = future::block_on(future::poll_once(&mut spawned)) {
-                    break result;
-                }
+                    results
+                };
 
-                self.executor.try_tick();
+                // Pin the futures on the stack.
+                pin!(fut);
+
+                // SAFETY: This function blocks until all futures complete, so we do not read/write the
+                // data from futures outside of the 'scope lifetime. However, rust has no way of knowing
+                // this so we must convert to 'static here to appease the compiler as it is unable to
+                // validate safety.
+                let fut: Pin<&mut (dyn Future<Output = Vec<T>>)> = fut;
+                let fut: Pin<&'static mut (dyn Future<Output = Vec<T>> + 'static)> =
+                    unsafe { mem::transmute(fut) };
+
+                // The thread that calls scope() will participate in driving tasks in the pool forward
+                // until the tasks that are spawned by this scope() call complete. (If the caller of scope()
+                // happens to be a thread in this thread pool, and we only have one thread in the pool, then
+                // simply calling future::block_on(spawned) would deadlock.)
+                let mut spawned = local_executor.spawn(fut);
```

Comment:
It looks like we're running all of the tasks on the local executor, regardless of the context they were spawned with. Maybe we should just have two scopes for simplicity / to avoid potential overhead when only one context is needed? Ex: `pool.scope()` and `pool.local_scope()`

Reply:
Ah wait. We're _awaiting_ (and i assume that means polling) using the local executor only. Which _might_ be fine? I'm not sure! Separate scopes would probably still be safest.

Reply:
Yep. The tasks run on the executor where they were spawned and are only awaited on the local executor. Doing it that way makes the code significantly cleaner because otherwise you have to basically duplicate every line because the executors have different types with no unifying trait.

I see that the tests I've added don't actually prove that this is running the tasks on the right threads. Let me add a test that demonstrates that. If you're still uncomfortable with it I can move to awaiting two groups of futures since I agree that is more obvious.

Reply:
I'm less worried about the parallelism (spawning the task with the proper executor is proof enough for me). I'm more worried about using the local executor to poll tasks that it didn't spawn. That seems a bit weird / might produce undefined behavior. I'm not sure how that's implemented (and even if the current implementation works, can we be sure that will always be the case?)

Reply:
That's a valid concern. Let go read more in async-executor/std::future and see what the guarantees actually are.

Reply:
Ok, I am convinced this is a reasonable thing to do because of the way async-task is designed. When you spawn a future via their API they produce two structs, a Task and a Runnable. The Task, which is the only part we interact with directly, is used only to check for completion and retrieve the result, it doesn't participate in polling the underlying future at all. The Runnable, which is retained by the Executor/LocalExecutor is what actually polls the future when `try_tick` is called. So using the Task in a composite future that waits for results is exactly what it is intended for. Turning the crank on the executor to actually run the main futures is a completely separate process.

Reply:
Alrighty I'm decently convinced at this point. Lets get this merged!